SecondBridge/SecondBridge/Data/BinarySearchTree.swift

```diff
@@ -0,0 +1,236 @@
+                return BinarySearchTree.Node(value, left: left, right: right.add(element))
+            } else {
+                return BinarySearchTree.Node(value, left: left.add(element), right: right)
+            }
+        }
+    }
+    
+    /**
+    Checks if a certain element is included in the current BST.
+    
+    - parameter element: The element to be checked.
+    
+    - returns: True if the BST contains this element.
+
+    */
+    func search(element:T) -> Bool{
+        
+        switch self{
+        case .Empty:
+            return false
+        case let .Node(value, _, _) where element == value:
+            return true
+        case let .Node(value, left, _) where value > element:
+            return left.search(element)
+        case let .Node(value, _, right) where value < element:
+            return right.search(element)
+        default:
+            print("Error")
+        }
+        return false
+    }
+    
+    /**
+    *  Removes the provided element from the current BST, and returns a new BST without that element
+    */
+    func remove(element:T) -> BinarySearchTree? {
+        
+        switch self {
+        case .Empty:
+            return self
+        case let .Node(value, left, right) where value > element:
+            return BinarySearchTree.Node(value, left: left.remove(element)!, right: right)
+        case let .Node(value, left, right) where value < element:
+            return BinarySearchTree.Node(value, left:left, right: right.remove(element)!)
+        case let .Node(value, .Empty, .Empty) where value == element:
+            return BinarySearchTree.Empty
+        case let .Node(value, left, .Empty) where value == element:
+            return left
+        case let .Node(value, .Empty, right) where value == element:
+            return right
+        case let .Node(value, left, right) where value == element:
+            let a = findMin(right)
+            switch a {
+            case let .Node(value, _,_):
+                return BinarySearchTree.Node(value, left: left, right: right.remove(value)!)
+            default:
+                return .Empty
+            }
+        default:
+            print("Error")
+            
+        }
+        return self
+        
+    }
+    
+    func findMin( root:BinarySearchTree) -> BinarySearchTree<T> {
+        
+        switch root{
+        case .Node(_, .Empty, _):
+            return root
+        case let .Node(_, left,_):
+            return findMin(left)
+        default:
+            return .Empty
+        }
+    }
+    
+    // MARK: - Printable
+    
+    var description: String {
+        switch self {
+        case .Empty:
+            return "."
+        case let .Node(value, left, right):
+            return "[\(left) \(value) \(right)]"
+        }
+    }
+    
+}
+
+// Traversal
+
+extension BinarySearchTree  {
+    
+    /*
+    * Returns a sorted array of node items.
+    */
+    func inOrderTraversal() -> [T] {
```

Comment:
I think that we should leave the name because is particular of Binary Search Tree. Also, I'll include two new  functions in this block, "preOrderTraversal" and "postOrderTraversal".